lib/internal/mime.js

```diff
@@ -0,0 +1,318 @@
+
+const NotHTTPTokenCodePoint = /[^!#$%&'*+\-.^_`|~A-Za-z0-9]/g;
+const NotHTTPQuotedStringCodePoint = /[^\t\u0020-~\u0080-\u00FF]/g;
+NotHTTPQuotedStringCodePoint[SymbolReplace] = null;
+
+const HTTP_TAB_OR_SPACE = '\t ';
+const HTTP_WHITESPACE = ArrayPrototypeJoin(['\r\n', ...HTTP_TAB_OR_SPACE], '');
+const BEGINNING_WHITESPACE = new RegExp(
+  `^[${StringPrototypeReplace(HTTP_WHITESPACE, /\W/g, '\\$&')}]*`);
+const ENDING_WHITESPACE = new RegExp(
+  `[${StringPrototypeReplace(HTTP_WHITESPACE, /\W/g, '\\$&')}]*$`);
+
+function toASCIILower(str) {
+  return StringPrototypeReplace(str, /[A-Z]/g, (c) => StringPrototypeToLowerCase(c));
+}
+
+function parseTypeAndSubtype(str) {
+  // Skip only HTTP whitespace from start
+  str = StringPrototypeReplace(str, BEGINNING_WHITESPACE, '');
+  let position = 0;
+  // read until '/'
+  const typeMatch = execPattern(/^[^/]*/, str);
+  if (!typeMatch ||
+    typeMatch[0] === '' ||
+    testPattern(NotHTTPTokenCodePoint, typeMatch[0])) {
+    throw new TypeError('Invalid MIME type');
+  }
+  position = typeMatch[0].length;
+  // skip '/'
+  position += 1;
+  const type = toASCIILower(typeMatch[0]);
+  // read until ';'
+  const subtypeMatch = execPattern(
+    /^[^;]*/,
+    StringPrototypeSlice(str, position));
+  if (subtypeMatch) {
+    position += subtypeMatch[0].length;
+    // skip ';'
+    position += 1;
+    subtypeMatch[0] = StringPrototypeReplace(
+      subtypeMatch[0],
+      ENDING_WHITESPACE,
+      '');
+  }
+  if (!subtypeMatch ||
+    subtypeMatch[0] === '' ||
+    testPattern(NotHTTPTokenCodePoint, subtypeMatch[0])) {
+    throw new TypeError('Invalid MIME subtype');
+  }
+  const subtype = toASCIILower(subtypeMatch[0]);
+  const target = ObjectCreate(null);
+  target.type = type;
+  target.subtype = subtype;
+  target.parametersString = StringPrototypeSlice(str, position);
+  return target;
+}
+
+function parseParametersString(str) {
+  const paramsMap = new Map();
+  str = StringPrototypeReplace(str, ENDING_WHITESPACE, '');
+  let position = 0;
+  while (position < str.length) {
+    // Skip any whitespace before parameter
+    const skip = execPattern(
+      BEGINNING_WHITESPACE,
+      StringPrototypeSlice(str, position))[0];
+    if (skip) position += skip.length;
+    // Read until ';' or '='
+    const parameterMatch = execPattern(
+      /^[^;=]*/,
+      StringPrototypeSlice(str, position));
+    const parameterString = toASCIILower(parameterMatch[0]);
+    position += parameterString.length;
+    // If we found a terminating character
+    if (position < str.length) {
+      // Safe to use because we never do special actions for surrogate pairs
+      const char = StringPrototypeCharAt(str, position);
+      // Skip the terminating character
+      position += 1;
+      // Ignore parameters without values
+      if (char === ';') {
+        continue;
+      }
+    }
+    // If we are at end of the string, it cannot have a value
+    if (position >= str.length) break;
```

Comment:
This can be just `else`, isn't it?
```js
    if (position < str.length) {
      // Safe to use because we never do special actions for surrogate pairs
      const char = StringPrototypeCharAt(str, position);
      // Skip the terminating character
      position += 1;
      // Ignore parameters without values
      if (char === ';') {
        continue;
      }
    } else {
      // If we are at end of the string, it cannot have a value.
      break;
    }
```

Reply:
doing it this way keeps base condition action and the reaction closer together, putting it in an else would require searching for the condition by scrolling up to the top of an if. this also allows us to avoid repetitive indentation for an `if` if the other is an abrupt jump like a `break`